internal/status/httproutestatus.go

```diff
@@ -0,0 +1,151 @@
+
+const ReasonNotImplemented RouteReasonType = "NotImplemented"
+const ReasonPathMatchType RouteReasonType = "PathMatchType"
+const ReasonForwardTo RouteReasonType = "ForwardTo"
+const ReasonValid RouteReasonType = "Valid"
+const ReasonErrorsExist RouteReasonType = "ErrorsExist"
+
+type HTTPRouteUpdate struct {
+	Fullname           types.NamespacedName
+	Conditions         []metav1.Condition
+	ExistingConditions []metav1.Condition
+	GatewayRef         types.NamespacedName
+	Generation         int64
+	TransitionTime     v1.Time
+}
+
+// AddCondition returns a metav1.Condition for a given ConditionType.
+func (routeUpdate *HTTPRouteUpdate) AddCondition(cond gatewayapi_v1alpha1.RouteConditionType, status metav1.ConditionStatus, reason RouteReasonType, message string) metav1.Condition {
+	newDc := metav1.Condition{
+		Reason:             string(reason),
+		Status:             status,
+		Type:               string(cond),
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		ObservedGeneration: routeUpdate.Generation,
+	}
+	routeUpdate.Conditions = append(routeUpdate.Conditions, newDc)
+	return newDc
+}
+
+// HTTPRouteAccessor returns a HTTPRouteUpdate that allows a client to build up a list of
+// v1.Conditions as well as a function to commit the change back to the cache when everything
+// is done. The commit function pattern is used so that the HTTPRouteUpdate does not need
+// to know anything the cache internals.
+func (c *Cache) HTTPRouteAccessor(route *gatewayapi_v1alpha1.HTTPRoute) (*HTTPRouteUpdate, func()) {
+	pu := &HTTPRouteUpdate{
+		Fullname:           k8s.NamespacedNameOf(route),
+		Conditions:         []metav1.Condition{},
+		ExistingConditions: c.getGatewayConditions(route.Status.Gateways),
+		GatewayRef:         c.gatewayRef,
+		Generation:         route.Generation,
+	}
+
+	return pu, func() {
+		c.commitHTTPRoute(pu)
+	}
+}
+
+func (c *Cache) commitHTTPRoute(pu *HTTPRouteUpdate) {
+	if len(pu.Conditions) == 0 {
+		return
+	}
+	c.httpRouteUpdates[pu.Fullname] = pu
+}
+
+func (routeUpdate *HTTPRouteUpdate) Mutate(obj interface{}) interface{} {
+	o, ok := obj.(*gatewayapi_v1alpha1.HTTPRoute)
+	if !ok {
+		panic(fmt.Sprintf("Unsupported %T object %s/%s in HTTPRouteUpdate status mutator",
+			obj, routeUpdate.Fullname.Namespace, routeUpdate.Fullname.Name,
+		))
+	}
+
+	httpRoute := o.DeepCopy()
+
+	var gatewayStatuses []gatewayapi_v1alpha1.RouteGatewayStatus
+	var conditionsToWrite []metav1.Condition
+
+	for _, cond := range routeUpdate.Conditions {
+
+		// Look through the newly calculated conditions, if they already exist
+		// on the object or if our obervation is stale, then leave them be,
+		// otherwise add them to the list to be written back to the API.
+		for _, existingCond := range routeUpdate.ExistingConditions {
+			if (cond.Status == existingCond.Status &&
+				cond.Reason == existingCond.Reason &&
+				cond.Message == existingCond.Message &&
+				cond.Type == existingCond.Type) ||
```

Comment:
Should this be an `&&` instead of an `||`? If not, them I'm not quite understanding the logic here.

Reply:
My logic was that if the object is the same, skip it. Or if the generation is out of date, skip it.